File: lectures/lecture-02-what-a-harness-actually-is/exercises/exercise-01-subsystem-auditor/solution/python/main.py
"""subsystem-auditor exercise, Python solution.

Audits repository trees for the minimal artifact of each harness subsystem.
Every signal is a language-neutral file: the audit criteria (SPEC.md) never
mention Python or TypeScript, and the TypeScript track reads the same trees.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

SUBSYSTEMS = ("instructions", "tools", "environment", "state", "feedback")


def _finding(present: bool, evidence: str | None) -> dict:
    return {"present": present, "evidence": evidence if present else None}


def audit_instructions(repo: Path) -> dict:
    for name in ("AGENTS.md", "CLAUDE.md"):
        path = repo / name
        if path.is_file() and path.read_text(encoding="utf-8").strip():
            return _finding(True, name)
    return _finding(False, None)


def audit_tools(repo: Path) -> dict:
    if (repo / "verify.sh").is_file():
        return _finding(True, "verify.sh")
    return _finding(False, None)


def audit_environment(repo: Path) -> dict:
    if (repo / "pyproject.toml").is_file() and (repo / ".python-version").is_file():
        return _finding(True, "pyproject.toml + .python-version")
    if (repo / "package.json").is_file() and (repo / ".nvmrc").is_file():
        return _finding(True, "package.json + .nvmrc")
    return _finding(False, None)


def audit_state(repo: Path) -> dict:
    if (repo / "feature_list.json").is_file() and (repo / "claude-progress.md").is_file():
        return _finding(True, "feature_list.json + claude-progress.md")
    return _finding(False, None)


def audit_feedback(repo: Path) -> dict:
    for name in ("AGENTS.md", "CLAUDE.md"):
        path = repo / name
        if path.is_file():
            for line in path.read_text(encoding="utf-8").split("\n"):
                if line.strip().startswith("- Verification:"):
                    return _finding(True, f"Verification line in {name}")
    return _finding(False, None)


AUDITS = {
    "instructions": audit_instructions,
    "tools": audit_tools,
    "environment": audit_environment,
    "state": audit_state,
    "feedback": audit_feedback,
}


def audit_repo(repo: Path) -> dict:
    subsystems = {name: AUDITS[name](repo) for name in SUBSYSTEMS}
    missing = [name for name in SUBSYSTEMS if not subsystems[name]["present"]]
    present = len(SUBSYSTEMS) - len(missing)
    return {
        "name": repo.name,
        "subsystems": subsystems,
        "score": f"{present}/{len(SUBSYSTEMS)}",
        "missing": missing,
    }


def main(argv: list[str]) -> int:
    if len(argv) != 2:
        print("usage: main.py <repos-dir>", file=sys.stderr)
        return 2
    repos_dir = Path(argv[1])
    if not repos_dir.is_dir():
        print(f"error: not a directory: {repos_dir}", file=sys.stderr)
        return 2
    repos = sorted(path for path in repos_dir.iterdir() if path.is_dir())
    report = {"repos": [audit_repo(repo) for repo in repos], "audited": len(repos)}
    print(json.dumps(report, indent=2))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: lectures/lecture-02-what-a-harness-actually-is/exercises/exercise-01-subsystem-auditor/solution/typescript/main.ts
// subsystem-auditor exercise, TypeScript solution.
//
// Audits repository trees for the minimal artifact of each harness
// subsystem. Every signal is a language-neutral file: the audit criteria
// (SPEC.md) never mention Python or TypeScript, and the Python track reads
// the same trees.

import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { join } from "node:path";
import { pathToFileURL } from "node:url";

const SUBSYSTEMS = ["instructions", "tools", "environment", "state", "feedback"] as const;

interface Finding {
  readonly present: boolean;
  readonly evidence: string | null;
}

function finding(present: boolean, evidence: string | null): Finding {
  return { present, evidence: present ? evidence : null };
}

function fileExists(repo: string, name: string): boolean {
  return existsSync(join(repo, name)) && statSync(join(repo, name)).isFile();
}

function auditInstructions(repo: string): Finding {
  for (const name of ["AGENTS.md", "CLAUDE.md"]) {
    if (fileExists(repo, name) && readFileSync(join(repo, name), "utf8").trim()) {
      return finding(true, name);
    }
  }
  return finding(false, null);
}

function auditTools(repo: string): Finding {
  return fileExists(repo, "verify.sh") ? finding(true, "verify.sh") : finding(false, null);
}

function auditEnvironment(repo: string): Finding {
  if (fileExists(repo, "pyproject.toml") && fileExists(repo, ".python-version")) {
    return finding(true, "pyproject.toml + .python-version");
  }
  if (fileExists(repo, "package.json") && fileExists(repo, ".nvmrc")) {
    return finding(true, "package.json + .nvmrc");
  }
  return finding(false, null);
}

function auditState(repo: string): Finding {
  if (fileExists(repo, "feature_list.json") && fileExists(repo, "claude-progress.md")) {
    return finding(true, "feature_list.json + claude-progress.md");
  }
  return finding(false, null);
}

function auditFeedback(repo: string): Finding {
  for (const name of ["AGENTS.md", "CLAUDE.md"]) {
    if (!fileExists(repo, name)) continue;
    const lines = readFileSync(join(repo, name), "utf8").split(/\r?\n/);
    if (lines.some((line) => line.trim().startsWith("- Verification:"))) {
      return finding(true, `Verification line in ${name}`);
    }
  }
  return finding(false, null);
}

const AUDITS: Record<string, (repo: string) => Finding> = {
  instructions: auditInstructions,
  tools: auditTools,
  environment: auditEnvironment,
  state: auditState,
  feedback: auditFeedback,
};

export function auditRepo(repo: string, name: string): Record<string, unknown> {
  const subsystems: Record<string, Finding> = {};
  for (const subsystem of SUBSYSTEMS) {
    const auditor = AUDITS[subsystem];
    if (auditor) subsystems[subsystem] = auditor(repo);
  }
  const missing = SUBSYSTEMS.filter((s) => !subsystems[s]?.present);
  const present = SUBSYSTEMS.length - missing.length;
  return {
    name,
    subsystems,
    score: `${present}/${SUBSYSTEMS.length}`,
    missing,
  };
}

function main(argv: readonly string[]): number {
  const reposDir = argv[2];
  if (!reposDir || argv.length !== 3) {
    console.error("usage: main.ts <repos-dir>");
    return 2;
  }
  if (!existsSync(reposDir) || !statSync(reposDir).isDirectory()) {
    console.error(`error: not a directory: ${reposDir}`);
    return 2;
  }
  const repos = readdirSync(reposDir)
    .filter((entry) => statSync(join(reposDir, entry)).isDirectory())
    .sort();
  const report = {
    repos: repos.map((name) => auditRepo(join(reposDir, name), name)),
    audited: repos.length,
  };
  process.stdout.write(JSON.stringify(report, null, 2) + "\n");
  return 0;
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) {
  process.exit(main(process.argv));
}
